### PR TITLE
Compatibility with GHC-8.0

### LINKED ISSUE #4: 8.0.1 compatibility
Would be nice to have 8.0.1 compatibility for this package.

### PR DESCRIPTION
**Pull request by sjakobi**
Fixes #4.

### CODE REVIEW

File: src/Present.hs [hunk 1]
```diff
@@ -132,6 +133,12 @@ normalizeType = go
             TH.StarT -> fail "Star (*) is not supported."
             TH.ConstraintT -> fail "Constraints are not supported."
             TH.LitT _ -> fail "Type-level literals are not supported."
+#if MIN_VERSION_template_haskell(2,11,0)
+            TH.InfixT{} -> fail "Infix type constructors are not supported."
+            TH.UInfixT{} -> fail "Unresolved infix type constructors are not supported."
+            TH.ParensT _ -> fail "Parenthesized types are not supported."
+            TH.WildCardT -> fail "Wildcard types are not supported."
```
Comment (sjakobi) [inline src/Present.hs]: Are the `fail`s here supposed to result in `Left`s?

File: src/Present.hs [hunk 2]
```diff
@@ -355,6 +370,20 @@ makeConstructor =
       ((\x y -> [x,y]) <$> makeSlot t1 <*> makeSlot t2)
     (TH.ForallC _ _ con) ->
       makeConstructor con
+#if MIN_VERSION_template_haskell(2,11,0)
+    TH.GadtC names slots _type ->
+      case names of
+        name:_ -> -- FIXME: What about the other names?
```
Comment (sjakobi) [inline src/Present.hs]: A cleaner solution might be to change the return type of `makeConstructor` to `Either String [Constructor]`. What do you think, @chrisdone?

**EDIT**: This is legal:

```
data T where
    A, B, C :: T
```

**Comment by sjakobi** [comment]
So, I haven't really tested this… If there are any specific cases that you'd like to see tested, please let me know!

How do I load that `tests.ghci` file with `stack ghci` BTW?

Oh, and while you're at it: Would you mind enabling Travis for this repo? I'll make a .travis.yml then.

File: src/Present.hs (hunk 2 above)
Comment (sjakobi) [reply src/Present.hs]: Oh, and how do I actually use present to get the various constructors listed?

File: src/Present.hs (hunk 1 above)
Comment (chrisdone) [reply src/Present.hs]: Yep, they're Left's.

File: src/Present.hs (hunk 2 above)
Comment (chrisdone) [reply src/Present.hs]: Indeed, I'd have it return a list. That'll require a `concat` elsewhere in the code, no biggie.

> Oh, and how do I actually use present to get the various constructors listed?

Not sure, check the revdeps of `makeConstructor`?

File: src/Present.hs (hunk 1 above)
Comment (sjakobi) [reply src/Present.hs]: Actually they are simple `error`s, but I can fix that in a follow-up PR.

**Comment by chrisdone** [comment]
> How do I load that tests.ghci file with stack ghci BTW?

`:script foo` should do it.

> Would you mind enabling Travis for this repo? I'll make a .travis.yml then.

Sure.

**Comment by chrisdone** [comment]
https://travis-ci.org/chrisdone/present

File: src/Present.hs (hunk 1 above)
Comment (chrisdone) [reply src/Present.hs]: They're left's:

screenshot 2017-03-16 10 03 40

But `presentType` calls `error` here (fail in Q monad): 

https://github.com/chrisdone/present/blob/5b00026940c4359cbef78139c6a917b1d3193172/src/Present.hs#L1069..L1071

Comment (sjakobi) [reply src/Present.hs]: > They're left's:

I'm not convinced! ;)

```
λ> fail "Bla" :: Either String Int
*** Exception: Bla
```

Also, see http://hackage.haskell.org/package/base-4.9.0.0/docs/Data-Either.html#v:fail

Comment (chrisdone) [reply src/Present.hs]: Oh, news to me, you can change them to `Left`s.

File: src/Present.hs (hunk 2 above)
Comment (sjakobi) [reply src/Present.hs]: > > Oh, and how do I actually use present to get the various constructors listed?
>
> Not sure, check the revdeps of makeConstructor?

I have opened https://github.com/chrisdone/present/issues/6 to remind me to look into this.

File: src/Present.hs (hunk 1 above)
Comment (sjakobi) [reply src/Present.hs]: > you can change them to Lefts.

I'll do that in a separate commit.

**Comment by sjakobi** [comment]
I have adjusted the type (and name) of `makeConstructor` and added a Travis config. Can you take another look?

File: src/Present.hs (hunk 1 above)
Comment (sjakobi) [reply src/Present.hs]: > I'll do that in a separate commit.

Scratch that, I've added 5a7ff87 to this PR.

**Comment by chrisdone** [comment]
Looks good. 👍

**Comment by sjakobi** [comment]
Merged! Can you make a release and add present back to Stackage?

**Comment by sjakobi** [comment]
> Merged! Can you make a release and add present back to Stackage?

Alternatively you could add me as a maintainer on Hackage and I'll do the release myself.

**Comment by chrisdone** [comment]
Added as maintainer!

**Comment by sjakobi** [comment]
> Merged! Can you make a release and add present back to Stackage?

I nearly forgot about the second part of that sentence: https://github.com/fpco/stackage/pull/2406

**Pull request merged**